### DISCUSSION TITLE
Insufficient version of system glibc (GLIBCXX_3.4.26 not found) - install works, but all tests fail

### DISCUSSION THREAD

**Discussion by WulfHans** [idaholab/moose#18378] [Q&A Getting Started] [answered]
I installed MOOSE exactly as described and the compilation worked without errors. However, all tests failed with an error message:
`/usr/lib/x86_64-linux-gnu/libstdc++.so.6: version GLIBCXX_3.4.26 not found`

Analysis of the error is simple by:
```
ldd ./lib/libmoose_test-opt.so.0
strings /usr/lib/x86_64-linux-gnu/libstdc++.so.6 | grep GLIBCXX
```
This shows easily, that the error message is just correct: libmoose links to a mixture of local libs (from miniconda) and system-wide libs. The systemwide glibc is indeed just a bit too old (it supports all up to GLIBCXX_3.4.25). For completeness: My system is Debian10, gcc 8.3.0, glibc 2.28

Some quick research shows, that such issues are not unheard of in context of conda. There are a lots of proposed solutions, including some rather scary ones (like upgrading your whole system-wide glibc) or annoying ones (like applying patchelf to a lot of things). 

My solution was inserting this into ~/miniconda3/envs/moose/etc/conda/activate.d/env_vars.sh:
```
export OLD_LD_LIBRARY_PATH=${LD_LIBRARY_PATH}
export LD_LIBRARY_PATH=~/miniconda3/envs/moose/lib:${LD_LIBRARY_PATH}
```
and this into the corresponding ~/miniconda3/envs/moose/etc/conda/deactivate.d/env_vars.sh:
```
export LD_LIBRARY_PATH=${OLD_LD_LIBRARY_PATH}
unset OLD_LD_LIBRARY_PATH
```

The miniconda environment contains its own (newer) libstdc++ and now the linker prefers these libs.
**This solves the main issue.** All tests pass now. However, any call is accompagnied by this:
`/bin/bash: /home/hans/miniconda3/envs/moose/lib/libtinfo.so.6: no version information available (required by /bin/bash)`

**My questions are:**
- **Is this supposed to happen?** Perhaps I am spoilt by how easy installing software usually is nowadays - but I suppose this should work out-of-the-box? What went wrong?
- **Is my fix the recommended solution?**
- **Are there any pitfalls to be aware of with this?**

**Comment by milljm** [comment, marked as answer]
There was a recent failure in dependency resolution for our moose-libmesh package. Can you verify what version of moose-$$$ packages are installed?

```bash
conda list | grep moose
```

I suspect (and hope), that you actually have an older moose-libmesh (or moose-tools) than what is supposed to be the correct combination of the two. The fix went in only an hour or so ago _manually_ into our Idaholab channel.

A good list:
```pre
moose-libmesh             2021.06.30              build_1    idaholab
moose-libmesh-vtk         6.3.0                   build_6    idaholab
moose-mpich               3.3.2                   build_6    idaholab
moose-petsc               3.15.1                  build_1    idaholab
moose-tools               2021.07.13       py39h3dbc78d_0    idaholab
```

What I think you might see for the error you are receiving:

Bad list (recorded from another post, hence the python differences):
```pre
moose-libmesh             2021.06.30              build_1    idaholab
moose-libmesh-vtk         6.3.0                   build_6    idaholab
moose-mpich               3.3.2                   build_6    idaholab
moose-petsc               3.15.1                  build_1    idaholab
moose-tools               2020.04.13       py37h734910c_0    idaholab
```

The fix was to moose-libmesh. Even though they look identical. The problem was with the internals of said package depending on a version of libnetcdf that the latest moose-tools was requesting. I simply fixed moose-libmesh specifically, and overwrote it on our Anaconda Idaholab channel (hence the identical name/version/build). In retrospect I probably should have bumped a version or build number. But I wanted to create the exact build our CI was going to build later (and it still hasn't). The fix is detailed here if you are curious: https://github.com/idaholab/moose/pull/18268

**Reply by WulfHans** [reply]
```
(moose) hans@hpc:~/Arbeit/moose_projects/moose/test$ conda list | grep moose
# packages in environment at /home/hans/miniconda3/envs/moose:
moose-libmesh             2021.06.30              build_0    idaholab
moose-libmesh-vtk         6.3.0                   build_5    idaholab
moose-mpich               3.3.2                   build_5    idaholab
moose-petsc               3.15.1                  build_0    idaholab
moose-tools               2020.04.13       py37h734910c_0    idaholab
```
I clearly have the bad list. However, its nice to see, that a fix is already on the way. Can I just update and revert the changes to the activate/deactive scripts?

**Reply by milljm** [reply]
The good news is the fix should already be available to you :). I would delete that pesky moose environment and recreate it. While you're at it, something we are working to move to is `mamba`. It's wicked quick at Conda solve/install times (all threaded operations). My condolences, if you were using `conda create -n moose moose-libmesh moose-tools` during this broken 'moose-libmesh' package. The solve time was horrendous.

Please try this next!!:

```bash
conda deactivate
conda env remove -n moose
conda install mamba
mamba create -n moose moose-libmesh moose-tools
conda activate moose
```

`mamba` is a drop in replacement for conda. So you can basically use all the same commands with mamba as you would with conda (`mamba install`, `mamba search`, `mamba remove`, etc). Except: `conda activate`. That one remains a necessity.

**Reply by WulfHans** [reply]
I did this and it seems, that I have the good list now:
```
(moose) hans@hpc:~/Arbeit/moose_projects/moose/test$ conda list | grep moose
# packages in environment at /home/hans/miniconda3/envs/moose:
moose-libmesh             2021.06.30              build_1    idaholab
moose-libmesh-vtk         6.3.0                   build_6    idaholab
moose-mpich               3.3.2                   build_6    idaholab
moose-petsc               3.15.1                  build_1    idaholab
moose-tools               2021.07.13       py39h3dbc78d_0    idaholab
```

However, i get the same error again:
`/home/hans/Arbeit/moose_projects/moose/test/moose_test-opt: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.26' not found (required by /home/hans/Arbeit/moose_projects/moose/test/lib/libmoose_test-opt.so.0)
`
I suppose I could fix it again by above mentioned small scripts, but if you have some further ideas, I would prefer trying those first.

**Reply by hugary1995** [reply]
It's very possible you have stale libs lying around after updating your conda environment.

You can try `make clobberall` to clean the moose directory and recompile. If that doesn't work, you can try `git clean -xfd` to clean all untracked files (remember to commit your work before doing so) then recompile.

**Reply by milljm** [reply]
There is also a new submodule in play which contains the pyhit parser: moosetools. While I believe `make clobberall` reaches this submodule, there is a sure-fire way to clean everything as @hugary1995  mentions with `git clean`. In order to get submodules clean I use:

```bash
cd moose
git submodule forearch --recursive git clean -xfd
git clean -xfd
```
The first git command removes any untracked files in all submodules, recursively.
The last git command removes any untracked files in the top-level repo. I always forget to run this one...

**Reply by WulfHans** [reply]
The last one finally solved it! Thanks for the swift and focussed help @all!
